### PR TITLE
Fix ArrayAccess limitations

### LINKED ISSUE #216: ArrayAccess limitations
@swaner mentioned an issue with the `ArrayAccess` casting found in the `ArrayCaster` class in #206. I have a fix for that specific issue, but in implementing it I have discovered some oversights in my original implementation of the `ArrayAccess` caster which causes it to not work for very basic `ArrayAccess` implementations that do not also implement `IteratorAggregate`.

My current thinking is to implement some validation in the `ArrayCaster` that ensures the object also implements `IteratorAggregate`. I don't think this would really be a breaking change, as it would simply add in a check for a scenario we couldn't cover. As a bonus, we would also gain support for `ArrayObject` implementations in that caster.

@brendt - Any thoughts here? Happy to create a PR if this sounds like the right course.

### PR DESCRIPTION
**Pull request by aidan-casey**
This PR fixes #216 and the limitations it describes.

### CODE REVIEW

**Comment by brendt** [comment]
Sorry it took so long, but this one looks good to merge! Can you rebase your branch? I've merged this one that's causing conflicts: https://github.com/spatie/data-transfer-object/pull/217

**Comment by aidan-casey** [comment]
@brendt Hmm... So it looks like that PR isn't accounting for key / values (noticing I did not account for that in the array caster of this PR either). Looks like I'm going to need to do some refactoring here, so it will probably be Friday or Saturday before I can get that done.

**Comment by brendt** [comment]
No problem :) Thanks for the heads-up!

**Comment by aidan-casey** [comment]
@brendt - Would appreciate a review on this. A couple highlights:

- Renamed `makeItem` to `castItem` to fit with the rest of the class.
- Added `mapInto` method which casts and maps items to their appropriate key in an `array` or `Traversable` (this deduplicates code and keeps support for key mapping in arrays).
- Reverted type hinting on `$value` from `array` to `mixed` because setting it to `array` was a breaking change. I thought about union types here, but since the entry variable is `mixed` I just kept that.
- Removed contents of the `ArrayCaster` from the README and just added a link.

If this gets merged, I recommend squashing commits because I forgot about the style fixer and got in a fight with it. 😅

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,10 @@ All notable changes to `data-transfer-object` will be documented in this file
 
 - Fix for an empty value being created when casting `ArrayAccess` objects (#216)
 - Add logic exception when attempting to cast `ArrayAccess` objects that are not traversable (#216)
+<<<<<<< HEAD
```
Comment (brendt) [inline CHANGELOG.md]: Careful, there's a merge conflict committed here

File: README.md [hunk 2]
```diff
@@ -176,8 +176,11 @@ abstract class BaseDataTransferObject extends DataTransferObject
 
 ### Using custom caster arguments
 
-Any caster can be passed custom arguments, the built-in `ArrayCaster` implementation is such an example:
+Any caster can be passed custom arguments, the built-in `ArrayCaster` [implementation](https://github.com/spatie/data-transfer-object/blob/master/src/Casters/ArrayCaster.php) is a good example of how this may be used.
 
+<<<<<<< HEAD
```
Comment (brendt) [inline README.md]: Same here

File: CHANGELOG.md (hunk 1 above)
Comment (aidan-casey) [reply CHANGELOG.md]: Yeah that was my bad (🤦🏻‍♂️), I caught it too late. Everything should be clean in the latest commit, so a squash and merge would clean up that history. Alternatively, I can clean up the Git history manually with a reset.

Comment (brendt) [reply CHANGELOG.md]: Don't worry about the history :)

Let me know when this is ready?

**Comment by aidan-casey** [comment]
@brendt - I think this is good to go.

**Comment by brendt** [comment]
Good!

**Comment by brendt** [comment]
Sorry for the late response btw, going to tag it now.

**Comment by brendt** [comment]
Tagged: https://github.com/spatie/data-transfer-object/releases/tag/3.4.0

**Pull request merged**